### PR TITLE
Support event listener through Hooks

### PR DESCRIPTION
**Pull request by FredLL-Avaiga**
Linked to [taipy-enterprise#362](https://github.com/Avaiga/taipy-enterprise/issues/362).

If a hook can handle this (hint enterprise with auth), it will.
Else nothing to do 

```
import datetime as dt
import typing as t

import pandas as pd

import taipy as tp
from taipy import Config, Frequency
from taipy.gui import get_state_id
from taipy.gui._hook import _Hook, _Hooks

data = pd.read_csv(
    "https://raw.githubusercontent.com/Avaiga/taipy-getting-started-core/develop/src/daily-min-temperatures.csv"
)


# Normal function used by Taipy
def predict(
    historical_temperature: pd.DataFrame, date_to_forecast: dt.datetime
) -> float:
    print(f"Running baseline...")
    historical_temperature["Date"] = pd.to_datetime(historical_temperature["Date"])
    historical_same_day = historical_temperature.loc[
        (historical_temperature["Date"].dt.day == date_to_forecast.day)
        & (historical_temperature["Date"].dt.month == date_to_forecast.month)
    ]
    return historical_same_day["Temp"].mean()


# Configuration of Data Nodes
historical_temperature_cfg = Config.configure_data_node("historical_temperature")
date_to_forecast_cfg = Config.configure_data_node("date_to_forecast")
predictions_cfg = Config.configure_data_node("predictions", storage_type="json")

# Configuration of tasks
predictions_cfg = Config.configure_task(
    "predict",
    predict,
    [historical_temperature_cfg, date_to_forecast_cfg],
    predictions_cfg,
)

# Configuration of scenario
scenario_cfg = Config.configure_scenario(
    id="my_scenario", task_configs=[predictions_cfg], frequency=Frequency.MONTHLY
)

if __name__ == "__main__":
    # Run of the Core
    tp.Orchestrator().run()

    class ListenerHook(_Hook):
        method_names = ["_init", "_add_event_listener", "_fire_event"]

        def __init__(self):
            super().__init__()
            self.listeners: t.Dict[str, list] = {}

        def _init(self, gui: tp.Gui):
            self.__gui = gui

        def _add_event_listener(
            self,
            event_name: str,
            listener: t.Callable[[str, t.Dict[str, t.Any]], None],
            with_state: t.Optional[bool] = False,
        ):
            self.listeners.setdefault(event_name, []).append(listener)

        def _fire_event(
            self, event_name: str, client_id: t.Optional[str] = None, payload: t.Optional[t.Dict[str, t.Any]] = None
        ):
            for func in self.listeners.setdefault(event_name, []):
                if client_id:
                    self.__gui._call_function_with_state(func, [client_id, payload or {}])
                else:
                    func(event_name, client_id, payload)

    _Hooks()._register_hook(ListenerHook())

    # Creation of the scenario and execution
    scenario = tp.create_scenario(scenario_cfg)
    scenario.historical_temperature.write(data)
    scenario.date_to_forecast.write(dt.datetime.now())
    tp.submit(scenario)

    scenario_md = """
<|{scenario}|scenario_selector|>

<|{scenario}|scenario|>

<|fire auth|button|on_action={lambda s: s.get_gui()._fire_event("authorization", get_state_id(s), {"override": "Who wants to create a scenario, anyway ?"})}|>
"""

    gui = tp.Gui(scenario_md)
    gui.run(title="1964 'Delete' and 'Promote to Primary' buttons always inactive")

```

### CODE REVIEW

File: taipy/gui/gui.py [hunk 1]
```diff
@@ -2907,3 +2907,44 @@ def set_favicon(self, favicon_path: t.Union[str, Path], state: t.Optional[State]
...
+        event_name: str,
+        listener: t.Union[
+            t.Callable[[str, t.Dict[str, t.Any]], None], t.Callable[[State, str, t.Dict[str, t.Any]], None]
+        ],
+        with_state: t.Optional[bool] = False,
+    ):
+        try:
+            _Hooks()._add_event_listener(event_name, listener, with_state)
+        except Exception as e:
+            _warn("Hooks: ", e)
```
Comment (FabienLelaquais) [inline taipy/gui/gui.py]:
```suggestion
            _warn("Exception in Hook._add_event_listener(): ", e)
```

File: taipy/gui/gui.py [hunk 2]
```diff
@@ -2907,3 +2907,44 @@ def set_favicon(self, favicon_path: t.Union[str, Path], state: t.Optional[State]
...
+            this_sid = getattr(request, "sid", None)
+            request.sid = None  # type: ignore[attr-defined]
+
+        try:
+            with self.get_flask_app().app_context():
+                if client_id:
+                    setattr(g, Gui.__ARG_CLIENT_ID, client_id)
+                _Hooks()._fire_event(event_name, client_id, payload)
+        except Exception as e:
+            _warn("Hooks: ", e)
```
Comment (FabienLelaquais) [inline taipy/gui/gui.py]:
```suggestion
            _warn("Exception in Hook._fire_event(): ", e)
```

File: taipy/gui/gui.py [hunk 3]
```diff
@@ -2907,3 +2907,44 @@ def set_favicon(self, favicon_path: t.Union[str, Path], state: t.Optional[State]
...
+    ):
+        try:
+            _Hooks()._add_event_listener(event_name, listener, with_state)
+        except Exception as e:
+            _warn("Hooks: ", e)
+
+    def _fire_event(
+        self, event_name: str, client_id: t.Optional[str] = None, payload: t.Optional[t.Dict[str, t.Any]] = None
+    ):
+        Timer(interval=0.1,
```
Comment (FabienLelaquais) [inline taipy/gui/gui.py]: Plese tell me why the delay...

File: taipy/gui/gui.py [hunk 4]
```diff
@@ -2907,3 +2907,44 @@ def set_favicon(self, favicon_path: t.Union[str, Path], state: t.Optional[State]
...
+            function=self.__do_fire_event,
+            args=(event_name, client_id, payload),
+        ).start()
+
+    def __do_fire_event(
+        self, event_name: str, client_id: t.Optional[str] = None, payload: t.Optional[t.Dict[str, t.Any]] = None
+    ):
+        this_sid = None
+        if request:
```
Comment (FabienLelaquais) [inline taipy/gui/gui.py:2920]: Where did *that* come from?

File: taipy/gui/gui.py [hunk 5]
```diff
@@ -2907,3 +2907,44 @@ def set_favicon(self, favicon_path: t.Union[str, Path], state: t.Optional[State]
...
+
+        try:
+            with self.get_flask_app().app_context():
+                if client_id:
+                    setattr(g, Gui.__ARG_CLIENT_ID, client_id)
+                _Hooks()._fire_event(event_name, client_id, payload)
+        except Exception as e:
+            _warn("Hooks: ", e)
+        finally:
+            if this_sid and request:
```
Comment (FabienLelaquais) [inline taipy/gui/gui.py]:
```suggestion
            if this_sid:
```
Or did I missed something?

**Review by FabienLelaquais — DISMISSED** [review]
👌

File: taipy/gui/_hook.py [hunk 6]
```diff
@@ -42,7 +42,7 @@ def _resolve_hook(*args, **kwargs):
                     func = getattr(hook, name)
                     if not callable(func):
                         raise Exception(f"'{name}' hook is not callable")
-                    res = getattr(hook, name)(*args, **kwargs)
+                    res = func(*args, **kwargs)
```
Comment (dinhlongviolin1) [inline taipy/gui/_hook.py:45]: thanks!

File: taipy/gui/gui.py (hunk 4 above)
Comment (dinhlongviolin1) [reply taipy/gui/gui.py:2920]: this comes from flask request. The context is there when a ws or http request is being handled

**Review by dinhlongviolin1 — DISMISSED** [review]

File: taipy/gui/gui.py (hunk 1 above)
Comment (FredLL-Avaiga) [reply taipy/gui/gui.py]: In fact you were right from the start (as usual ?)
All the exceptions are caught and logged in the _Hooks class.
I'll remove the try/catch

File: taipy/gui/gui.py (hunk 3 above)
Comment (FredLL-Avaiga) [reply taipy/gui/gui.py]: # fire event is executed after 0.1s to let the current state manipulation finish
        # (that would change the credentials for example)
        # in ts, we would use Promise.resolve(() => {__do_fire_event})

File: taipy/gui/gui.py (hunk 4 above)
Comment (FredLL-Avaiga) [reply taipy/gui/gui.py:2920]: That's what we do when we emulate a request to get a state

File: taipy/gui/gui.py (hunk 5 above)
Comment (FredLL-Avaiga) [reply taipy/gui/gui.py]: I hope not

**Review by FabienLelaquais — DISMISSED** [review]
👍

**Comment by FredLL-Avaiga** [comment]
```py
import inspect
from inspect import isclass, ismodule, isroutine


def a_function():
  pass

async def a_async_function():
  pass

class a_class():
  def a_method(self):
    pass

  @staticmethod
  def a_static_method():
    pass

class a_callable_class():
  def __call__(self):
    print("in call")
    pass


def get_status(a_callable):
  return (f"{a_callable.__name__ if hasattr(a_callable, '__name__') else str(type(a_callable))}"
          f" is {'callable' if callable(a_callable) else 'not callable'}"
          f" and is {'a routine' if isroutine(a_callable) else 'not a routine'}"
          f" and is {'a class' if isclass(a_callable) else 'not a class'}")


print(get_status(a_function))
print(get_status(a_class))
print(get_status(a_class.a_method))
print(get_status(a_callable_class))
print(get_status(a_callable_class()))
print(get_status(inspect))
print(get_status(a_async_function))
print(get_status(print))

```

returns 
```
a_function is callable and is a routine and is not a class
a_class is callable and is not a routine and is a class
a_method is callable and is a routine and is not a class
a_callable_class is callable and is not a routine and is a class
<class '__main__.a_callable_class'> is callable and is not a routine and is not a class
inspect is not callable and is not a routine and is not a class
a_async_function is callable and is a routine and is not a class
print is callable and is a routine and is not a class
```

File: taipy/gui/utils/callable.py [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+# Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+# an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+# specific language governing permissions and limitations under the License.
+
+import typing as t
+from inspect import isclass
+from types import LambdaType
+
+
+def _is_function_like(s: t.Any) -> bool:
```
Comment (FabienLelaquais) [inline taipy/gui/utils/callable.py]: I would have called this  '_is_function' to be honest.
... and... can a class be callabled? Seriously??

File: taipy/gui/utils/callable.py [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+    return callable(s) and not isclass(s)
+
+
+def _get_function_like_name(s: t.Any) -> str:
```
Comment (FabienLelaquais) [inline taipy/gui/utils/callable.py]:
```suggestion
def _get_function_name(s: t.Any) -> str:
```
or even `_function_name()`

**Review by FabienLelaquais — DISMISSED** [review]
👌

File: taipy/gui/utils/callable.py (hunk 7 above)
Comment (FredLL-Avaiga) [reply taipy/gui/utils/callable.py]: Why not
of course, that's why I switched to isroutine

**Review by FabienLelaquais — APPROVED** [review]
👍

**Pull request merged**